java-client/barrage/src/main/java/io/deephaven/client/impl/BarrageSubscriptionImpl.java

```diff
@@ -127,66 +141,74 @@ public void onNext(final BarrageMessage barrageMessage) {
+            }
+
+            log.error().append(BarrageSubscriptionImpl.this).append(": unexpectedly closed by other host").endl();
+            resultTable.handleBarrageError(new RequestCancelledException("Barrage subscription closed by server"));
+            cleanup();
         }
     }
 
     @Override
-    public boolean isCompleted() {
-        return completed;
+    public Future<Table> entireTable() {
+        return partialTable(null, null, false);
     }
 
     @Override
-    public BarrageTable entireTable() throws InterruptedException {
-        return entireTable(true);
+    public Future<Table> partialTable(RowSet viewport, BitSet columns) {
+        return partialTable(viewport, columns, false);
     }
 
     @Override
-    public BarrageTable entireTable(boolean blockUntilComplete) throws InterruptedException {
-        return partialTable(null, null, false, blockUntilComplete);
+    public Future<Table> snapshotEntireTable() {
+        return snapshotPartialTable(null, null, false);
     }
 
     @Override
-    public BarrageTable partialTable(RowSet viewport, BitSet columns) throws InterruptedException {
-        return partialTable(viewport, columns, false, true);
+    public Future<Table> snapshotPartialTable(RowSet viewport, BitSet columns) {
+        return snapshotPartialTable(viewport, columns, false);
     }
 
     @Override
-    public BarrageTable partialTable(RowSet viewport, BitSet columns, boolean reverseViewport)
-            throws InterruptedException {
-        return partialTable(viewport, columns, reverseViewport, true);
+    public Future<Table> snapshotPartialTable(RowSet viewport, BitSet columns, boolean reverseViewport) {
+        isSnapshot = true;
+        return partialTable(viewport, columns, reverseViewport);
     }
 
     @Override
-    public BarrageTable partialTable(RowSet viewport, BitSet columns, boolean reverseViewport,
-            boolean blockUntilComplete) throws InterruptedException {
+    public Future<Table> partialTable(RowSet viewport, BitSet columns, boolean reverseViewport) {
         synchronized (this) {
-            if (!connected) {
-                throw new UncheckedDeephavenException(
-                        this + " is no longer an active subscription and cannot be retained further");
+            if (!isConnected()) {
+                throw new UncheckedDeephavenException(this + " is no longer connected and cannot be retained further");
             }
             if (subscribed) {
-                throw new UncheckedDeephavenException(
-                        "BarrageSubscription objects cannot be reused.");
+                throw new UncheckedDeephavenException("Barrage subscription objects cannot be reused");
             }
             subscribed = true;
         }
 
+        if (isSnapshot) {
+            future = new CompletableFutureAdapter();
+        } else {
+            future = new UpdateGraphAwareFutureAdapter(resultTable.getUpdateGraph());
+        }
```

Comment:
I'm concerned we have lurking NPEs since we delay this initialization. :-(

Reply:
The future is read in the vp-change callback which cannot be invoked until after the `observer.onNext` (which happens below). This looks safe to me. Note that the future is volatile due to the var's need to cross thread boundaries.

We cannot create this much earlier as we do not know whether the user wants a snapshot or a subscription until right here (after checking that the object is not-yet-used and still connected).